torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,199 @@
+
+// see [value refinement algorithm]
+
+struct IntegerValueRefiner {
+  IntegerValueRefiner(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    if (!checkForPossibleRefinements(graph_->block())) {
+      return false;
+    }
+    ListRefinement refinements;
+    RefineListLens(graph_->block(), refinements);
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          auto non_const_index = 1 - const_index;
+          if (n->inputs().at(const_index)->node()->kind() == prim::Constant &&
+              n->inputs().at(non_const_index)->uses().size() > 1) {
+            return true;
+          }
+        }
+      }
+      for (Block* block : n->blocks()) {
+        if (checkForPossibleRefinements(block)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  ListRefinement RefineListLens(Block* b, ListRefinement block_refinements) {
+    active_refinements_.push_back(&block_refinements);
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          if (auto ival = constant_as<int64_t>(n->inputs().at(const_index))) {
+            IntegerRefinement refine;
+            refine[n->inputs().at(1 - const_index)] = *ival;
+            info_[n->output()] = n->kind() == aten::eq
+                ? BooleanRefinementMapping::TrueRefinements(std::move(refine))
+                : BooleanRefinementMapping::FalseRefinements(std::move(refine));
+          }
+        }
+      }
+      if (n->kind() == prim::RaiseException) {
+        throwing_blocks_.insert(b);
+      }
+
+      for (size_t input = 0; input < n->inputs().size(); ++input) {
+        Value* input_v = n->inputs().at(input);
+        if (!input_v->type()->cast<IntType>()) {
+          continue;
+        }
+
+        if (auto refine = tryFindRefinement(input_v)) {
+          WithInsertPoint guard(n);
+          auto refine_constant =
+              graph_->insertConstant(static_cast<int64_t>(*refine));
+          n->replaceInputWith(input_v, refine_constant);
+          changed_ = true;
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = info_.count(if_n.cond()) != 0;
+        ListRefinement empty;
+        auto true_block_refinements = RefineListLens(
+            if_n.thenBlock(),
+            has_cond_ref ? info_[if_n.cond()].true_refine() : empty);
+        auto false_block_refinements = RefineListLens(
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+
+        // we are looking for cases where we can replace
```

Comment:
I feel carving this special handling out into a separate function would help the readability of the algorithm.